### PR TITLE
Fix: Unutilize fee for revolving period with optional periods

### PR DESCRIPTION
**Pull request by eitanfr**

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeEnumerations.java [hunk 1]
```diff
@@ -170,7 +170,7 @@ public static EnumOptionData chargeCalculationType(final ChargeCalculationType t
             break;
             case PERCENT_OF_UNUTILIZED_AMOUNT:
                 optionData = new EnumOptionData(ChargeCalculationType.PERCENT_OF_UNUTILIZED_AMOUNT.getValue().longValue(),
-                        ChargeCalculationType.PERCENT_OF_UNUTILIZED_AMOUNT.getCode(), "% Unutilized Amount");
+                        ChargeCalculationType.PERCENT_OF_UNUTILIZED_AMOUNT.getCode(), "% annually of Unutilized Amount");
```
Comment (davidyaha) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeEnumerations.java:173]: Maybe "Unutilized Amount (Annual %)"

Comment (eitanfr) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeEnumerations.java:173]: starting with "%" is a fineract convention,
then should I change all enum types?

Comment (davidyaha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/charge/service/ChargeEnumerations.java:173]: okay so maybe "% Daily Unutilized Amount (Annual %)". This is quite confusing but it is actually what's going on here...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 2]
```diff
@@ -2694,26 +2701,97 @@ private BigDecimal getExpectedDisbursedAmount(LocalDate untilDate) {
...
+
+        BigDecimal chargeAmount = BigDecimal.ZERO;
+        LocalDate fromCalcDate = periodStart;
+        // For calculating unutilized amount disbursements are negative and transactions are positive
+        List<TransactionHelper> transactionsAndDisbursements =  Stream.concat(disbursementsInPeriodStream, transactionInPeriodStream)
+                .sorted(Comparator.comparing(TransactionHelper::getDate))
+                .collect(Collectors.toList());
+
+        // Add last date that the fee will be calculated for.
+        var revolvingEndDate = Optional.ofNullable(revolvingPeriodEndDate).map(date -> LocalDate.ofInstant(date.toInstant(), ZoneId.systemDefault()));
```
Comment (davidyaha) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: why use `var` and not `LocalDate`?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 3]
```diff
@@ -2694,26 +2701,97 @@ private BigDecimal getExpectedDisbursedAmount(LocalDate untilDate) {
...
+        LocalDate lastDateCalculatedFor = this.getLoanProduct().isRevolving() && revolvingEndDate.isPresent() &&
+                                        revolvingEndDate.get().isBefore(periodEnd) ? revolvingEndDate.get() : periodEnd;
+        transactionsAndDisbursements.add(new TransactionHelper(lastDateCalculatedFor, BigDecimal.ZERO));
+
+        for (TransactionHelper transactionOrDisbursement :transactionsAndDisbursements) {
```
Comment (davidyaha) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: I think you need to run `spotlessApply`

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractLoanScheduleGenerator.java [hunk 4]
```diff
@@ -2101,61 +2096,109 @@ private BigDecimal getDisbursedAmount(final LoanApplicationTerms loanApplication
         return interestCharges;
     }
 
-    private BigDecimal calcUnutilizedAmount(final LoanApplicationTerms loanApplicationTerms, LocalDate untilDate,
-            boolean inRevolvingPeriodOnly, List<LoanTransaction> transactions) {
-        BigDecimal disbursedAmount = getDisbursedAmount(loanApplicationTerms, untilDate);
-        Money approvedPrincipal = loanApplicationTerms.getApprovedPrincipal();
+    private boolean checkDateBetween(LocalDate dateToCheck, LocalDate startDate, LocalDate endDate) {
```
Comment (davidyaha) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractLoanScheduleGenerator.java]: Already seen this method on Loan... Why not add it to CalendarUtils?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractLoanScheduleGenerator.java [hunk 5]
```diff
@@ -2101,61 +2096,109 @@ private BigDecimal getDisbursedAmount(final LoanApplicationTerms loanApplication
...
+        return (startDate == null || !dateToCheck.isBefore(startDate)) && (endDate == null || !dateToCheck.isAfter(endDate));
+    }
+
+    private BigDecimal calcUntilizedChargeAmount(LoanApplicationTerms applicationTerms, final LoanCharge loanCharge,
```
Comment (davidyaha) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractLoanScheduleGenerator.java]: there's a reason why this is implemented both in Loan and here?

**Pull request merged**